### PR TITLE
feat: HID Usages + Keys Standardisation

### LINKED ISSUE #21: Standardize Keys
- Revisit/refactor current `dt-bindings/zmk/keys.h`
- Include any useful values for USB or HOGP from HID Usage Tables v1.12:
https://www.usb.org/sites/default/files/documents/hut1_12v2.pdf
- Consider values and naming conventions found in existing KB firmware, Linux, Windows, Android, etc.
- Decide upon any conventions for ZMK.
- Open up consultation with the ZMK contributor community before finalizing.

### LINKED ISSUE #158: Document scancode preprocessor definitions
Currently the preprocessor defs for use in keymaps is a bit opaque at times, some abbreviations aren't obvious as to what they are supposed to be. Even end-of-line comments in `keys.h` would be of use.

### LINKED ISSUE #217: feat(HID): Preprocessor definitions for HID Usage Tables 1.21
Prerequisite of #21.

- Specification: https://usb.org/sites/default/files/hut1_2.pdf
- Mirror specification where possible.
- Include pages and usages.
- Develop appropriate naming conventions.
- Aim for self-documenting where possible.
- Replace existing definitions.

### LINKED ISSUE #234: feat(HID): enhance reports to support higher usage codes
Facilitates the efforts of #21.
- Enhance keyboard reports up to *at least* `0xFB`.
- Enhance consumer reports to support double byte values.

### LINKED ISSUE #246: bug(keys):  &kp above 0x65 does not work
I'm pretty confident that none of these current `&kp` key codes/names work for *anyone*.

```
#define UNDO 0x7A
#define CUT 0x7B
#define COPY 0x7C
#define PSTE 0x7D

#define CRRT 0xC3
#define PRCT 0xC4
#define LABT 0xC5
#define RABT 0xC6
#define AMPS 0xC7
#define PIPE 0xC9
#define COLN 0xCB
#define HASH 0xCC
#define KSPC 0xCD
#define ATSN 0xCE
#define BANG 0xCF

#define VOLU 0x80
#define VOLD 0x81
```

### Rationale
- ZMK's `hid` system currently guards against (blocks) anything above `GUI 0x65`.
- The NKRO keyboard report currently only has enough space for 13 bytes or 104 codes.
- The default USB HID endpoint max packet size (`CONFIG_HID_INTERRUPT_EP_MPS`) is only 16 bytes which would block anything higher than 0x70.
- Regardless of the above issues, my research suggests that only these codes are likely to have support by an OS:
```
#define UNDO 0x7A
#define CUT 0x7B
#define COPY 0x7C
#define PSTE 0x7D

#define VOLU 0x80
#define VOLD 0x81
```

### Moving Forward
Most - if not all - of these codes are going to be superseded by the keys standardization (#21) and possibly by the upcoming shift/modifiers enhancement (#86).

In the short-term, we need to make it clear to everyone - board/shield developers and users - that ...
- these codes don't currently work
- these key names should not be used in shields etc. (especially incoming PRs)
- some of these key names are about to be deprecated, some superseded (by #21).

### End-to-End Testing
What's a tad surprising is that this isn't a new sets of bugs - they've been around since at least June.  These key codes have been incorporated into user keymaps and these shield keymaps:
- `corne`
- `iris`
- `lily58`
- `sofie`
- `qaz`
 
... which suggests our testing and QA needs improvement.
 
IMO, I think it highlights that we cannot just rely on developer or user feedback.  So I suggest we look into developing verifiable end-to-end testing - from ZMK to OS - which can be run by board/shield developers.  My ad-hoc end-to-end system that I threw together for developing #21 has already paid for itself - timewise - and this issue tops the cake.  I'd suggest we explore the possibilities for this - probably using Docker or some multiplatform approach.

### PR DESCRIPTION
**Pull request by innovaker**
- Adds HID Usages for ZMK (#217).
- Standardizes key aliases/codes (#21 - still open for RFC).
- Enhances HID reports to take advantage of the new codes (#234).

### Issues
- Closes #217.
- Closes #21.
- Closes #234.
- Supersedes and closes #246.
- Supersedes and closes #158.
- Supersedes and closes #161.

### Testing
- End-to-end tests covering:
  - all key aliases
  - both the keyboard (keypad) and consumer HID reports
- ZMK on Nordic's nRF52840 Dongle.
  - `kscan-mock` running on dongle systematically pressed and released each key.
- Node scripts on Raspberry Pi (Raspbian Buster)
   - Scripts running on the Pi parsed events from `/dev/hidrawX` and `/dev/input/...` devices and compared incoming data against expected key usage codes.
   - Scripts not included in PR because end-to-end framework isn't generic enough and some of the code is my own.
- Test logs:
  - [e2e-2020-10-04T12-14-37-usb.log](https://github.com/zmkfirmware/zmk/files/5323138/e2e-2020-10-04T12-14-37-usb.log)
  - [e2e-2020-10-04T12-14-37-ble.log](https://github.com/zmkfirmware/zmk/files/5323139/e2e-2020-10-04T12-14-37-ble.log)
- Once HID reports were enhanced, usage codes themselves never caused any errors.
- Untested usage codes should also be viable for the future as everything was developed systematically via scripting.

### CODE REVIEW

**Comment by innovaker** [comment]
Regarding HID usage codes, the keen observers will notice that I included all the usages - even the ones we don't utilize yet.

This is because:
a) it keeps doors open for the future (I personally plan on using some of them myself).
b) it wasn't much extra work.

I can remove them and refresh the PR if people really don't want them included for whatever reason.

**Comment by innovaker** [comment]
I suggest we give this PR at least a few days to settle because there may be some developers that missed the RFC for #21.  It would encourage multi-platform testing too.

**Comment by innovaker** [comment]
@tominabox1 has encountered unintended consequences of 2da9078ac43f7e38205b7a231e54df1520862ac7 when users have defined their own behaviour instances: https://discordapp.com/channels/719497620560543766/719544230497878116/762294497081163806
i.e. in their keymaps

I think it's still worth deprecating `USAGE_KEYPAD` and `USAGE_CONSUMER` for their replacements (see `hid_usage.h` to see what I mean).  But should we:
- ask our more advanced users to update their keymaps on merge of this issue?
- provides legacy fallback defines - i.e. "symlinks" in `hid_usage_legacy.h` - for the short term?

Preference on this @petejohanson please?

For anyone trying to test, in the meantime the replacements are:
`USAGE_KEYPAD`: `HID_USAGE_KEY`
`USAGE_CONSUMER`: `HID_USAGE_CONSUMER`

**Comment by innovaker** [comment]
Follow-up to my last comment ... it's mainly a concern for anyone who has merged https://github.com/zmkfirmware/zmk/pull/153 into their local as it uses:
- `USAGE_KEYPAD`: 20 times
- `USAGE_CONSUMER`: 6 times

**Comment by innovaker** [comment]
I've force-pushed a new version that adds legacy values for `USAGE_KEYPAD` and `USAGE_CONSUMER`.  @tominabox1, could you give it a try please (please remember to strip out the previous pull)?

**Comment by innovaker** [comment]
@tominabox1 has gracefully retested.  There are a few merge conflicts with #153.  So if you're experimenting with @okke-formsma's modifiers branch and are looking to test it alongside this PR, then you'll need to resolve those merge conflicts.  The force push earlier reintroduced `USAGE_KEYPAD` and `USAGE_CONSUMER` as legacy values for this purpose.

**Comment by innovaker** [comment]
Most recent test logs from the re-push:
- [e2e-2020-10-04T15-43-02-ble.log](https://github.com/zmkfirmware/zmk/files/5323556/e2e-2020-10-04T15-43-02-ble.log)
- [e2e-2020-10-04T15-43-02-usb.log](https://github.com/zmkfirmware/zmk/files/5323557/e2e-2020-10-04T15-43-02-usb.log)

**Comment by innovaker** [comment]
As I develop the associated documentation, I'm starting to notice small areas for improvement.  So I'll probably re-push some changes after pushing the first draft of #218 (coming soon).

**Review by petejohanson — COMMENTED** [review]
This really looks solid. Looking forward to the updated version for final review.

**Comment by innovaker** [comment]
### Updates
- Updated usages to HID Usage Tables 1.21
  - See https://usb.org/sites/default/files/hut1_21.pdf.
- Deprecated codes found in bug #246.
- Added `K_` convention.
  - See https://github.com/zmkfirmware/zmk/issues/21#issuecomment-706690476.
- Added extra codes discovered by scanning `0x000` to `0xFFF` through `evtest` (both pages).
- Minor improvements to various names.
- Integrated RFC improvements from @petejohanson and @innovaker.
  - See discussions in https://docs.google.com/spreadsheets/d/1s6URDIL7pqcMiUvY8s0esWvLIA1dKNmHqp6jEUtrqTw/.
- Refactored legacy names into inline deprecated names with inline warnings.
- Replaced deprecated names in code, tests and documentation.
- Refactored names in code and tests with fully-qualified names.
- Reformatted example keymaps in documentation to better fit new key names.
- Reordered commits to be more logical and `git blame` friendly.
- Refactored commit messages for improved readability.

### Testing
- app
  - Same as before.
  - logs:
    - [e2e-2020-10-28T11-49-58-ble.log](https://github.com/zmkfirmware/zmk/files/5454672/e2e-2020-10-28T11-49-58-ble.log)
    - [e2e-2020-10-28T11-49-58-usb.log](https://github.com/zmkfirmware/zmk/files/5454673/e2e-2020-10-28T11-49-58-usb.log)
- docs: 
  - `eslint . --ext js,jsx,md,mdx`
  - `npx prettier --write .`

### Reviews / Feedback
- Please pull and test it, your existing keymap *should* still work!!!
- Any feedback or suggestions welcomed!
- There was a lot of data involved, can you spot any oversights?
- For reviewing the key names, please review in conjunction with PR #308 (it will make it easier!).

**Comment by tominabox1** [comment]
I've tried mapping the media keys (cp and kp varieties) and can't seem to get them to work on my laptop anymore. I remember we tested them a few weeks ago and they worked so I'm not sure what may have changed.

**Comment by innovaker** [comment]
Thanks @tominabox1.  You might also try the various other[ media key codes](https://5f99f3c599fd970007c1451a--zmk.netlify.app/docs/codes/media).  Before I settled on the names, I did cursory testing on applications within Android, Linux and Windows and discovered the support to be inconsistent and occasionally not what you might expect.  More on that here if you're curious: https://github.com/zmkfirmware/zmk/issues/21#issuecomment-702182102.

There could also be difference between what the OS supports and which applications support them.  I touched on that in this [message](https://discord.com/channels/719497620560543766/719544230497878116/771342122834067496) on Discord.

**Comment by adamcstephens** [comment]
Volume works for me across Linux and Mac. Will need to get some time trying some of the other media keys. Otherwise, I'm using this branch as my daily driver and will report any other issues.

File: app/include/zmk/hid.h
Comment (petejohanson) [inline app/include/zmk/hid.h:89]: In doing some further testing on this, when using my MDK kit over USB, I do *not* get key events with this change and the change to the `ZMK_HID_MAX_KEYCODE` value. My *guess* is that this is related to the USB data being too large after this change, and causing problems for the onboard USB HUB, perhaps?

@innovaker Thoughts on reverting just these changes, in favor of switching to a `6KRO` (and future `#KRO`) type HID report/descriptor for this in a future PR?

Doing so would of course limit the actual *use* of some of these new HID keycodes defined here, but allow us to move forward with other changes while we work on iterating on the actual transmitted HID report encoding/format.

**Comment by innovaker** [comment]
Heads up to anyone currently testing/exploring/using this PR!

Further to Pete's earlier feedback, today we replaced a core part of the system on which this PR relies.  It's currently undergoing testing as PR #313.

Once that's ready and merged, I'll then re-push this PR rebased on top `main`'s new `HEAD`, and we'll kindly ask you to try the refreshed version.  It'll likely be out sometime tomorrow or the weekend.

Thanks for testing it, we appreciate the feedback!

File: app/include/zmk/hid.h
Comment (innovaker) [reply app/include/zmk/hid.h:89]: Casual observers, this was discussed throughout the day on Discord.  We replaced a core part of the system in an effort to fix this issue.  More delays follow below.

**Comment by innovaker** [comment]
### Changes
- Rebased onto #313 and #269.
- Refactored HID pages into `hid_usage_pages.h`.
- Replaced Zephyr usages with ZMK usages in HID reports.
- Minor comments tweaks.

### Testing
- As before.
  - [e2e-2020-10-31T16-05-57-ble.log](https://github.com/zmkfirmware/zmk/files/5469394/e2e-2020-10-31T16-05-57-ble.log)
  - [e2e-2020-10-31T16-05-57-usb.log](https://github.com/zmkfirmware/zmk/files/5469395/e2e-2020-10-31T16-05-57-usb.log)

**Comment by innovaker** [comment]
@tominabox1, @adamcstephens, @okke-formsma, anyone else testing - should be good to go! :)

**Comment by okke-formsma** [comment]
lgtm

**Comment by innovaker** [comment]
Having now laid the groundwork for shifted keys (a future PR), a minor quandary presented itself for *this* issue that we didn't previously consider.

What should we call `'` and `"`?
 - `APOSTROPHE`/`APOS` and `QUOTE`?
 - `SINGLE_QUOTE`/`SQT` and `DOUBLE_QUOTE`/`DQT`?
 - `QUOTE`/`QUOTE` and `DOUBLE_QUOTE`/`DQUOTE`?
 - `QUOTE` and `QUOTE2`?
 - Should we use a mix?
 - Some other idea I haven't considered?

@petejohanson, any thoughts please?

**Comment by tominabox1** [comment]
I would prefer to see us use something similar to (if not the same as) QMK because I suspect it will be most familiar to new zmk users. They use "QUOTE" - so maybe we use "QUOTES" that way we cover "single" and "double" quotes.

**Comment by innovaker** [comment]
From what I've observed, it seems to be a toss up between:
 - `QUOTE`/`QT` and `QUOTES`/`QTS`
 - `SINGLE_QUOTE`/`SQT` and `DOUBLE_QUOTE(S)`/`DQT`

but also including `APOSTROPHE`/`APOS` for the Brits regardless.

@petejohanson, you favor the latter I believe?  If so, preference on plural vs. singular for the double please?

**Comment by innovaker** [comment]
With the shifted keys work complete (that's the second part of #86, *not* #153), if we are able to finalise their names soon then I'm considering just rolling their *names* into this PR - but *without* the modifier `LS()` wrappers until #153 has completed its review process.  So effectively they'd just an alias of their non-shifted key.

Why?  Because of #246.  This PR currently perpetuates the problems of #246 because they were incorrectly mapped in the original version.  We kept them in until now because we didn't have an alternative.  Now we do, so I'd rather get them in.  The legwork is done, we just need to agree on the naming scheme.

**Comment by innovaker** [comment]
### Changes
 - Added *shifted key name* candidates.
   - This effectively "corrects" the identities for the remainder of #246 (except `CURU` which is being removed).
   - Once modifiers #153 has completed its review cycle, the `LS(...)` wrappers will be added to the shifted key names.
   - The ongoing RFC for the shifted key names (**green** cells) can be found here:
     https://docs.google.com/spreadsheets/d/1s6URDIL7pqcMiUvY8s0esWvLIA1dKNmHqp6jEUtrqTw/

### Testing
 - As before.
   - [e2e-2020-11-01T14-01-31-ble.log](https://github.com/zmkfirmware/zmk/files/5470991/e2e-2020-11-01T14-01-31-ble.log)
   - [e2e-2020-11-01T14-01-31-usb.log](https://github.com/zmkfirmware/zmk/files/5470992/e2e-2020-11-01T14-01-31-usb.log)


### Notes
 - This could be the last push unless (either):
   - The shifted keys RFC introduces minor changes.
   - I come across any problems (unlikely) during my final self-review.
 - Testers, your efforts have been much appreciated!  You're welcome to retest this version if you'd like.  There shouldn't be any noticeable differences ... unless I screwed up 👍

**Comment by joelspadin** [comment]
My main concern with using `QUOTE` and `QUOTES` is how similar they are. It would be easy to make a typo and get the wrong one, and I'm not sure it's very clear what the difference is, especially given that on a standard keyboard the legend on the key is two different quote symbols and thus "quotes" in the sense of "multiple kinds of quote" instead of "double quote".

My vote is for making single vs. double explicit. Alternatively, something like `QUOTE` for `'` and `DOUBLE_QUOTE` for `"` would also avoid that confusion, but I don't think that's as consistent.

File: app/include/zmk/hid_usage_pages.h
Comment (innovaker) [inline app/include/zmk/hid_usage_pages.h:47]: I tossed up between a few conventions for the usage pages:
 - `HID_USAGE_XXXXX`
 - `HID_USAGE_PAGE_XXXXX`
 - `HID_USAGE_XXXXX_PAGE`

I went for the first because the form facilitated natural namespaces for the HID usage definitions (see `hid_usage.h`).

But having now put the pages into the HID reports, they can be a bit confusing.   So I'm honestly not sure which is best from a readability perspective. 

Any thoughts from anyone please?

**Comment by innovaker** [comment]
@tominabox1 , @adamcstephens, @okke-formsma.  Depending on what comes out of tonight (US time), it's looking like final testing will be tomorrow, probably towards the evening as people will be working.  Any changes between now and then should be minor though, unless something goes wrong.

There's no expectations of you, but I thought I'd give you the heads up nonetheless.  Thanks for all your support so far!

**Comment by tominabox1** [comment]
Testing a number of codes: 
International codes don't seem to do anything on my MacOS Catalina system (may not know how they work). 
`C_VOL_UP` works but `K_VOL_UP` does not (nor does `K_VOL_UP2`)
Interestingly, `C_MUTE` and `K_MUTE` works, but `K_MUTE2` does not
Eject keys don't trigger any behavior on my system
`CLEAR` and `PRIOR` don't seem to do anything
`C_ALT_AUDIO_INC` `C_PLAY` Don't do anything
 `C_PP` Works
`C_REC` doesn't trigger anything
 `K_PP` `C_PAUSE` `C_STOP` `K_STOP2` `K_STOP3` doesn't do anything
`C_STOP_EJECT`doesn't do anything
`C_NEXT` works
`K_NEXT` doesn't work
`C_PREV` works
`K_PREV` doesn't work
`C_FF` `C_RW` Works - same behavior as next/prev
`C_SHUFFLE` doesn't work
Also tested editing `cp` codes (cut/copy/paste) and did't see any behaviors on the host system.

**Comment by innovaker** [comment]
Thanks a lot @tominabox1!  That's roughly what I'd expect from mac OS given what I've picked up in passing along this journey.

**Comment by innovaker** [comment]
### Changes
- 3x shifted key names updated after RFC feedback.

### Testing
- As before.
  - [e2e-2020-11-02T09-41-02-ble.log](https://github.com/zmkfirmware/zmk/files/5474006/e2e-2020-11-02T09-41-02-ble.log)
  - [e2e-2020-11-02T09-41-02-usb.log](https://github.com/zmkfirmware/zmk/files/5474007/e2e-2020-11-02T09-41-02-usb.log)

**Comment by innovaker** [comment]
Please don't merge this issue until I sign it off.

File: app/include/zmk/hid_usage_pages.h
Comment (petejohanson) [reply app/include/zmk/hid_usage_pages.h:47]: I think I favor the way you have it here. I think it's plenty clear.

File: app/include/dt-bindings/zmk/keys.h
Comment (petejohanson) [inline app/include/dt-bindings/zmk/keys.h:280]: With all of these added "shifted codes", would the replacing these with the shifted varients once that's merged. e.g.:

```suggestion
#define DOUBLE_QUOTES (LS(HID_USAGE_KEY_KEYBOARD_APOSTROPHE_AND_QUOTE))
```

?

That's really my only outstanding item I would want clarity on. Otherwise this looks great.

**Review by petejohanson — COMMENTED** [review]
Just one question on the shifted codes.

File: app/include/dt-bindings/zmk/keys.h
Comment (CrossR) [reply app/include/dt-bindings/zmk/keys.h:280]: The rebased branch with modifiers and this PR has these keycodes with the shifted variant: https://github.com/innovaker/zmk/tree/hid-keys-modifiers-rebase. I.e. 

https://github.com/innovaker/zmk/blob/6e52af9870ddb88e05a948e4d9160911e6543e9f/app/include/dt-bindings/zmk/keys.h#L278-L280

So I expect that is the plan, though it was emphasised that branch is a "dry run" so there may be some other tiny bits of work around integrating them I guess.

(Caveat aside, I've been using that  branch (hid-keys-modifiers-rebase) all of today with no issues except a few expected collisions between names and my own defines when doing the build, which is expected. Made my keymap a lot shorter, I only need defines for UK keycodes now, as the names don't make much sense there).

Comment (petejohanson) [reply app/include/dt-bindings/zmk/keys.h:280]: @CrossR Indeed. Good catch. Thanks.

**Review by petejohanson — APPROVED** [review]
@innovaker This looks good to me. Ready to merge when you are.

**Comment by innovaker** [comment]
Thanks for looking through it.  I'll be doing the rebase today.

**Comment by innovaker** [comment]
### Changes
 - Rebased onto `main`'s `HEAD`.
 - Aligned modifier PR #153 to this PR's changes.
 - Added shifting wrappers `LS()` to shifted key definitions.

### Testing
- As before (same checklist), but also tested with the deprecated codes completely removed.
- end-2-end tests (including deprecated codes):
  - [e2e-2020-11-03T14-03-40-ble.log](https://github.com/zmkfirmware/zmk/files/5482146/e2e-2020-11-03T14-03-40-ble.log)
  - [e2e-2020-11-03T14-03-40-usb.log](https://github.com/zmkfirmware/zmk/files/5482147/e2e-2020-11-03T14-03-40-usb.log)
- end-2-end tests (without deprecated codes):
  - [e2e-2020-11-03T13-54-35-ble.log](https://github.com/zmkfirmware/zmk/files/5482142/e2e-2020-11-03T13-54-35-ble.log)
  - [e2e-2020-11-03T13-54-35-usb.log](https://github.com/zmkfirmware/zmk/files/5482143/e2e-2020-11-03T13-54-35-usb.log)

### Notes
 - This is the release candidate!
  So, it will not change unless either:
   - Testing uncovers problems (as it did last Friday).
   - I suddenly spot something important that I missed (my checklist was huge, so it's unlikely).
 - The rebase had some conflicts and this is a big PR, so I recommend:
   - Give it a day to "settle" before merging.
     - This should also give me enough time to prepare the documentation for simultaneous release which be a big bonus.
   - Testers *sign off* before it's merged.
 - I appreciate everyone is anxious to get this out.  I personally can't wait to dust my hands of it.  But *please* be patient!  We just need one more day for stability testing.

### Testers
 - Please *leave a comment* if you're intending to test this release candidate.  We need visibility of what's going on.

File: app/include/dt-bindings/zmk/keys.h
Comment (petejohanson) [inline app/include/dt-bindings/zmk/keys.h:8]: Actually just noticed this one *organiational* issue. In general, I've been keeping *all* headers that might be included into keymaps, and are by requirement *only* preprocessor directives, in the `include/dt-bindings/` directory so there is a clear delineation of "headers used in keymaps". Those headers of course can/will be used by the code, but it's clear they are used in keymaps this way.

@innovaker Any problems moving the couple files into those directories? Shouldn't require re-testing, beyond just verifying builds, and would help keep that distinction intact.

**Comment by innovaker** [comment]
Hold my beer, @petejohanson has requested an organisational change.  Will re-push in 2 hours.

File: app/include/dt-bindings/zmk/keys.h
Comment (innovaker) [reply app/include/dt-bindings/zmk/keys.h:8]: To confirm:
- `zmk/hid_usage.h` to `dt-bindings/zmk/hid_usage.h`
- `zmk/hid_usage_pages.h` to `dt-bindings/zmk/hid_usage_pages.h`
?

Comment (petejohanson) [reply app/include/dt-bindings/zmk/keys.h:8]: That's the only two I can see, yeah.

Comment (innovaker) [reply app/include/dt-bindings/zmk/keys.h:8]: Any preference on the `#include` orders for those?

Comment (innovaker) [reply app/include/dt-bindings/zmk/keys.h:8]: nevermind on my last

Comment (innovaker) [reply app/include/dt-bindings/zmk/keys.h:8]: Consider it done.

**Comment by innovaker** [comment]
Release Candidate ... 2?

### Changes
 - Moved `hid_usage.h` and `hid_usage_page.h` from `include/zmk` into `include/dt-bindings/zmk` as per @petejohanson's [request](https://github.com/zmkfirmware/zmk/pull/236#discussion_r516802721).

### Testing
 - As before.
 - end-to-end tests (including deprecated codes):
   - [e2e-2020-11-03T18-22-05-ble.log](https://github.com/zmkfirmware/zmk/files/5483252/e2e-2020-11-03T18-22-05-ble.log)
   - [e2e-2020-11-03T18-22-05-usb.log](https://github.com/zmkfirmware/zmk/files/5483253/e2e-2020-11-03T18-22-05-usb.log)

### Testers
 - Please *leave a comment* if you're intending to test this release.  Thanks!

**Comment by innovaker** [comment]
Release Candidate 3 ... the final?  Cause I'm off to bed.

### Changes
 - Rebased onto #325, because it's useful and 10 is a lucky number (this will be the last rebase!!!)

### Testing
 - As before.
 - end-to-end tests (including deprecated codes):
   - [e2e-2020-11-03T20-17-12-ble.log](https://github.com/zmkfirmware/zmk/files/5483848/e2e-2020-11-03T20-17-12-ble.log)
   - [e2e-2020-11-03T20-17-12-usb.log](https://github.com/zmkfirmware/zmk/files/5483849/e2e-2020-11-03T20-17-12-usb.log)

### Testers
 - Please *leave a comment* if you're intending to test this release.  Thanks!

**Comment by innovaker** [comment]
I believe @tominabox1 is doing some testing.

**Comment by adamcstephens** [comment]
This works for me. Thanks!

**Comment by petejohanson** [comment]
@tominabox1 has noted on Discord it was working fine for him.

**Comment by chenkevinh** [comment]
Seems to be working fine for me on Windows so far.

**Comment by petejohanson** [comment]
Also got some testing verification from @Nicell w/ his nice60, no reported problems. Merging this!

**Pull request merged**
